// rollup-plugin-css-only/dist/index.cjs.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var fs = require('fs');
var rollupPluginutils = require('rollup-pluginutils');
var path = _interopDefault(require('path'));

function css(options) {
  if ( options === void 0 ) options = {};

  var filter = rollupPluginutils.createFilter(options.include || ['**/*.css'], options.exclude);
  var styles = {};
  var dest = options.output;
  var changes = 0;

  return {
    name: 'css',
    transform: function transform(code, id) {
      if (!filter(id)) {
        return
      }

      // When output is disabled, the stylesheet is exported as a string
      if (options.output === false) {
        return {
          code: 'export default ' + JSON.stringify(code),
          map: { mappings: '' }
        }
      }

      // Keep track of every stylesheet
      // Check if it changed since last render
      if (styles[id] !== code && (styles[id] || code)) {
        styles[id] = code;
        changes++;
      }

      return ''
    },
    generateBundle: function generateBundle(opts, bundle) {
      // No stylesheet needed
      if (!changes || options.output === false) {
        return
      }
      changes = 0;

      // Combine all stylesheets
      var css = '';
      for (var id in styles) {
        css += styles[id] || '';
      }

      // Emit styles through callback
      if (typeof options.output === 'function') {
        options.output(css, styles, bundle);
        return
      }

      if (typeof dest !== 'string') {
        // Don't create unwanted empty stylesheets
        if (!css.length) {
          return
        }

        // Guess destination filename
        dest =
          opts.file ||
          (Array.isArray(opts.output)
            ? opts.output[0].file
            : opts.output && opts.output.file) ||
          opts.dest ||
          'bundle.js';
        if (dest.endsWith('.js')) {
          dest = dest.slice(0, -3);
        }
        dest = dest + '.css';
      }

      // Emit styles to file
      return new Promise(function (resolve, reject) {
        var ref = path.parse(dest);
        var dir = ref.dir;

        fs.promises
          .mkdir(dir, { recursive: true })
          .then(function () {
            fs.writeFile(dest, css, function (err) {
              if (err) {
                reject(err);
              } else {
                resolve();
              }
            });
          })
          .catch(function (err) {
            reject(err);
          });
      })
    }
  }
}

module.exports = css;
